File: client-android/lib/src/main/kotlin/stasis/client_android/lib/security/DefaultOAuthClient.kt
package stasis.client_android.lib.security

import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import stasis.client_android.lib.api.clients.exceptions.AccessDeniedFailure
import stasis.client_android.lib.api.clients.exceptions.EndpointFailure
import stasis.client_android.lib.api.clients.internal.ClientExtensions
import stasis.client_android.lib.security.HttpCredentials.Companion.withCredentials
import stasis.client_android.lib.utils.AsyncOps.async
import stasis.client_android.lib.utils.NonFatal.nonFatal
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Companion.flatMap
import stasis.client_android.lib.utils.Try.Failure
import stasis.client_android.lib.utils.Try.Success

class DefaultOAuthClient(
    private val tokenEndpoint: String,
    client: String,
    clientSecret: String
) : OAuthClient {
    private val credentials =
        HttpCredentials.BasicHttpCredentials(username = client, password = clientSecret)

    private val http: OkHttpClient = OkHttpClient()

    private val moshi: Moshi = Moshi.Builder()
        .add(KotlinJsonAdapterFactory())
        .build()

    override suspend fun token(
        scope: String?,
        parameters: OAuthClient.GrantParameters
    ): Try<AccessTokenResponse> {
        val scopeParams = scope?.let { mapOf("scope" to it) } ?: emptyMap()

        val grantParams = when (parameters) {
            is OAuthClient.GrantParameters.ClientCredentials -> mapOf(
                "grant_type" to OAuthClient.GrantType.ClientCredentials
            )
            is OAuthClient.GrantParameters.ResourceOwnerPasswordCredentials -> mapOf(
                "grant_type" to OAuthClient.GrantType.ResourceOwnerPasswordCredentials,
                "username" to parameters.username,
                "password" to parameters.password
            )
            is OAuthClient.GrantParameters.RefreshToken -> mapOf(
                "grant_type" to OAuthClient.GrantType.RefreshToken,
                "refresh_token" to parameters.refreshToken
            )
        }

        return try {
            val body = (scopeParams + grantParams).toList()
                .fold(FormBody.Builder()) { builder, (paramName, paramValue) ->
                    builder.add(name = paramName, value = paramValue)
                }.build()

            val response = jsonRequest<AccessTokenResponse> { builder ->
                builder
                    .url(tokenEndpoint)
                    .post(body = body)
            }

            response
        } catch (e: Throwable) {
            Failure(e.nonFatal())
        }
    }


    private suspend inline fun request(block: (Request.Builder) -> Request.Builder): Response {
        val request = block(Request.Builder().withCredentials(credentials)).build()
        return http.newCall(request).async()
    }

    private fun Response.successful(): Try<Response> = when {
        this.isSuccessful -> Success(this)
        this.code == ClientExtensions.StatusUnauthorized -> Failure(AccessDeniedFailure())
        else -> Failure(EndpointFailure("Server responded with [${this.code} - ${this.message}]"))
    }

    private suspend inline fun <reified T> jsonRequest(block: (Request.Builder) -> Request.Builder): Try<T> =
        request(block).successful().flatMap { response ->
            val result = withContext(Dispatchers.IO) {
                response.body?.use {
                    moshi.adapter(T::class.java).fromJson(it.source())
                }
            }

            when (result) {
                null -> Failure(EndpointFailure("Expected response data but none was found"))
                else -> Success(result)
            }
        }


    companion object {
        const val StatusUnauthorized: Int = 401
    }
}

File: client-ios/lib/Sources/StasisClientLib/Security/OAuth/DefaultOAuthClient.swift
import Foundation

public struct DefaultOAuthClient: OAuthClient {
    public static let statusUnauthorized: Int = 401

    private let tokenEndpoint: URL
    private let credentials: HttpCredentials
    private let transport: any HttpTransport

    public init(
        tokenEndpoint: String,
        client: String,
        clientSecret: String,
        transport: any HttpTransport = URLSession.shared
    ) throws {
        guard let url = URL(string: tokenEndpoint) else {
            throw EndpointFailure(message: "Invalid token endpoint URL [\(tokenEndpoint)]")
        }
        self.tokenEndpoint = url
        self.credentials = .basic(username: client, password: clientSecret)
        self.transport = transport
    }

    public func token(
        scope: String?,
        parameters: GrantParameters
    ) async -> Result<AccessTokenResponse, Error> {
        var items: [URLQueryItem] = []
        if let scope {
            items.append(URLQueryItem(name: "scope", value: scope))
        }
        switch parameters {
        case .clientCredentials:
            items.append(URLQueryItem(name: "grant_type", value: GrantType.clientCredentials))
        case let .resourceOwnerPasswordCredentials(username, password):
            items.append(URLQueryItem(name: "grant_type", value: GrantType.resourceOwnerPasswordCredentials))
            items.append(URLQueryItem(name: "username", value: username))
            items.append(URLQueryItem(name: "password", value: password))
        case let .refreshToken(token):
            items.append(URLQueryItem(name: "grant_type", value: GrantType.refreshToken))
            items.append(URLQueryItem(name: "refresh_token", value: token))
        }

        var components = URLComponents()
        components.queryItems = items

        do {
            var request = URLRequest(url: tokenEndpoint)
            request.httpMethod = "POST"
            request.setValue(
                "application/x-www-form-urlencoded",
                forHTTPHeaderField: "Content-Type"
            )
            request.httpBody = Data((components.percentEncodedQuery ?? "").utf8)
            let authorized = request.withCredentials(credentials)

            let (data, response) = try await transport.data(for: authorized)
            guard let http = response as? HTTPURLResponse else {
                return .failure(EndpointFailure(message: "Server responded with a non-HTTP response"))
            }
            switch http.statusCode {
            case 200..<300:
                let decoded = try JSONDecoder().decode(AccessTokenResponse.self, from: data)
                return .success(decoded)
            case Self.statusUnauthorized:
                return .failure(AccessDeniedFailure())
            default:
                let message = HTTPURLResponse.localizedString(forStatusCode: http.statusCode)
                return .failure(EndpointFailure(
                    message: "Server responded with [\(http.statusCode) - \(message)]"
                ))
            }
        } catch {
            return .failure(error)
        }
    }
}
